src/main/java/io/appform/dropwizard/sharding/BundleCommonBase.java

```diff
@@ -15,66 +14,117 @@
+import org.jasypt.encryption.pbe.StandardPBEByteEncryptor;
+import org.jasypt.encryption.pbe.StandardPBEStringEncryptor;
+import org.jasypt.hibernate5.encryptor.HibernatePBEEncryptorRegistry;
+import org.jasypt.iv.StringFixedIvGenerator;
+import org.reflections.Reflections;
 
 @Slf4j
 public abstract class BundleCommonBase<T extends Configuration> implements ConfiguredBundle<T> {
 
   protected final List<TransactionListener> listeners = new ArrayList<>();
   protected final List<TransactionFilter> filters = new ArrayList<>();
-
   protected final List<TransactionObserver> observers = new ArrayList<>();
 
-  protected final List<Class<?>> initialisedEntities;
+  protected final List<Class<?>> initialisedEntities = new ArrayList<>();
+  protected final Map<String, EntityMeta> initialisedEntitiesMeta = new HashMap<>();
 
-  protected final Map<String, EntityMeta> initialisedEntitiesMeta = Maps.newHashMap();
+  private final Lock entityInitialisationBlockedLock = new ReentrantLock();
+  private volatile boolean entityInitialisationBlocked = false;
 
   protected TransactionObserver rootObserver;
 
-  protected BundleCommonBase(Class<?> entity, Class<?>... entities) {
-    this.initialisedEntities = ImmutableList.<Class<?>>builder().add(entity).add(entities).build();
-    validateAndBuildEntitiesMeta(initialisedEntities);
+  protected BundleCommonBase(final Class<?> entity, final Class<?>... entities) {
+    initialiseEntitiesAndMeta(ImmutableList.<Class<?>>builder()
+            .add(entity).add(entities).build());
   }
 
-  protected BundleCommonBase(List<String> classPathPrefixList) {
-    Set<Class<?>> entities = new Reflections(classPathPrefixList).getTypesAnnotatedWith(
-        Entity.class);
+  protected BundleCommonBase(final List<String> classPathPrefixList) {
+    final var entities = new Reflections(classPathPrefixList).getTypesAnnotatedWith(Entity.class);
     Preconditions.checkArgument(!entities.isEmpty(),
-        String.format("No entity class found at %s",
-            String.join(",", classPathPrefixList)));
-    this.initialisedEntities = ImmutableList.<Class<?>>builder().addAll(entities).build();
-    validateAndBuildEntitiesMeta(initialisedEntities);
+            String.format("No entity class found at %s", String.join(",", classPathPrefixList)));
+    initialiseEntitiesAndMeta(entities);
+  }
+
+  public final void registerEntities(@NonNull final Class<?>... entities) {
+    entityInitialisationBlockedLock.lock();
+    try {
+      if (entityInitialisationBlocked) {
+        throw new UnsupportedOperationException("Entity registration is not supported after run method execution.");
+      }
+      initialisePendingInitialisationEntitiesAndMeta(Arrays.asList(entities));
+    } finally {
+      entityInitialisationBlockedLock.unlock();
+    }
+  }
+
+  public final void registerEntities(@NonNull final List<String> classPathPrefixList) {
+    entityInitialisationBlockedLock.lock();
```

Comment:
instead of taking lock in individual methods, we can take the same lock in `initialisePendingInitialisationEntitiesAndMeta`